Framework/Algorithms/src/CreateDetectorTable.cpp

```diff
@@ -0,0 +1,334 @@
+DECLARE_ALGORITHM(CreateDetectorTable)
+
+void CreateDetectorTable::init() {
+  declareProperty(std::make_unique<WorkspaceProperty<Workspace>>(
+                      "InputWorkspace", "", Direction::Input),
+                  "The name of the workspace to take as input.");
+
+  declareProperty(std::make_unique<ArrayProperty<int>>("WorkspaceIndices",
+                                                       Direction::Input),
+                  "If left empty then all workspace indices are used.");
+  setPropertySettings(
+      "WorkspaceIndices",
+      std::make_unique<EnabledWhenWorkspaceIsType<MatrixWorkspace>>(
+          "InputWorkspace", true));
+
+  declareProperty("IncludeData", false,
+                  "Include the first value from each spectrum.");
+  setPropertySettings(
+      "IncludeData",
+      std::make_unique<EnabledWhenWorkspaceIsType<MatrixWorkspace>>(
+          "InputWorkspace", true));
+
+  declareProperty(
+      std::make_unique<WorkspaceProperty<TableWorkspace>>(
+          "DetectorTableWorkspace", "", Direction::Output,
+          PropertyMode::Optional),
+      "The name of the outputted detector table workspace, if left empty then "
+      "the input workspace name + \"-Detectors\" is used.");
+}
+
+void CreateDetectorTable::exec() {
+  Workspace_sptr inputWS = getProperty("InputWorkspace");
+  bool include_data = getProperty("IncludeData");
+  std::vector<int> indices = getProperty("WorkspaceIndices");
+
+  ITableWorkspace_sptr detectorTable;
+  // Standard MatrixWorkspace
+  auto matrix = boost::dynamic_pointer_cast<MatrixWorkspace>(inputWS);
+  if (matrix) {
+    detectorTable =
+        createDetectorTableWorkspace(matrix, indices, include_data, g_log);
+
+    if (detectorTable == nullptr) {
+      throw std::runtime_error("The instrument has no sample.");
+    }
+  } else {
+    auto peaks = boost::dynamic_pointer_cast<IPeaksWorkspace>(inputWS);
+    if (peaks) {
+      detectorTable = peaks->createDetectorTable();
+    }
+  }
+
+  if (detectorTable == nullptr) {
+    throw std::runtime_error(
+        "Detector table can only be created for matrix and peaks workspaces.");
+  }
+
+  if (getPropertyValue("DetectorTableWorkspace") == "") {
+    setPropertyValue("DetectorTableWorkspace",
+                     inputWS->getName() + "-Detectors");
+  }
+
+  setProperty("DetectorTableWorkspace", detectorTable);
+}
+
+/**
+ * Create the instrument detector table workspace from a MatrixWorkspace
+ * @param ws :: A pointer to a MatrixWorkspace
+ * @param indices :: Limit the table to these workspace indices
+ * @param include_data :: If true then first value from the each spectrum is
+ * displayed
+ *
+ * @return A pointer to the table workspace of detector information
+ */
+ITableWorkspace_sptr
+createDetectorTableWorkspace(const MatrixWorkspace_sptr &ws,
+                             const std::vector<int> &indices,
+                             const bool &include_data, Logger &logger) {
```

Comment:
```suggestion
                             const bool include_data, Logger &logger) {
```

No need to pass `bool` (or any other POD types) by const reference.